game.py

```diff
@@ -0,0 +1,908 @@
+from math import inf
+from copy import copy, deepcopy
+import random
+import pickle
+import time
+
+class GameState():
+    
+    def __init__(self, board, player_next, connect_number, number_of_moves, moves_list = []):
+        self._board = board
+        self._player_next = player_next
+        self._connect_number = connect_number
+        self._number_of_moves = number_of_moves
+        self._moves_list = moves_list
+        # self._last_move = last_move
+        # self._previous_game_state = previous_game_state
+
+    def _copy(self):
+        board = deepcopy(self._board)
+        player_next = deepcopy(self._player_next)
+        connect_number = deepcopy(self._connect_number)
+        number_of_moves = deepcopy(self._number_of_moves)
+        moves_list = deepcopy(self._moves_list)
+        # previous_game_state = deepcopy(self)
+        return GameState(board, player_next, connect_number, number_of_moves, moves_list)
+                
+    def get_board(self):
+        return self._board
+    
+    def get_player_next(self):
+        return self._player_next
+
+    def get_last_move(self):
+        return self._moves_list[-1]
+    
+    def get_connect_number(self):
+        return self._connect_number
+
+    def get_number_of_moves(self):
+        return self._number_of_moves
+    
+    def make_move(self, move, player_next = True):
+               
+        self._board[move[0]][move[1]] = self._player_next
+        self._player_next = - self._player_next
+        self._moves_list.append(move)
+        self._number_of_moves = self._number_of_moves + 1
+
+        # return game_state
+        
+    def unmake_move(self):
+        move = self.get_last_move()
+        self._board[move[0]][move[1]] = None
+        self._player_next = - self._player_next
+        self._number_of_moves = self._number_of_moves - 1
+        self._moves_list = self._moves_list[: -1]

```

Comment:
Note this line isn't quite the inverse of the `list.append(x)` you've used above in `make_move()`. This line actually creates a new list to copy over all the existing contents of the list (excluding the final one). You may want to replace this with `list.pop(-1)` to be consistent.